### PR TITLE
Add iterate for Reverse{<:NamedTuple}

### PR DESCRIPTION
**Pull request by bkamins**
Fixes the problem mentioned in https://github.com/JuliaLang/julia/pull/42991#issuecomment-966190513

### CODE REVIEW

File: test/iterators.jl [hunk 1]
```diff
@@ -687,17 +687,19 @@ end
...
     @test collect(take(Iterators.reverse(cycle(1:3)), 7)) == collect(take(cycle(3:-1:1), 7))
     let r = repeated(3.14159)
         @test Iterators.reverse(r) === r
     end
-    let t = (2,3,5,7,11)
+    for t in [(), (1,), (2, 3, 5, 7, 11), NamedTuple(), (a=1,), (a=2, b=3, c=5, d=7, e=11)]
         @test Iterators.reverse(Iterators.reverse(t)) === t
         @test first(Iterators.reverse(t)) === last(t)
         @test last(Iterators.reverse(t)) === first(t)
+        @test collect(Iterators.reverse(t)) == [11, 7, 5, 3, 2]
```
Comment (bkamins) [inline test/iterators.jl:702]:
```suggestion
        @test collect(Iterators.reverse(t)) == reverse(collect(t))
```

**Comment by bkamins** [comment]
Also relatedly I would propose to come back to https://github.com/JuliaLang/julia/pull/16604#issuecomment-966380990 PR and make a decision.

File: test/iterators.jl [hunk 2]
```diff
@@ -687,17 +687,19 @@ end
...
+                pairs(50:60), zip(1:10,21:30,51:60), product(1:3, 10:12), repeated(3.14159, 5),
+                (a=2, b=3, c=5, d=7, e=11))
         @test squash(collect(Iterators.reverse(itr))) == reverse(squash(collect(itr)))
     end
     @test collect(take(Iterators.reverse(cycle(1:3)), 7)) == collect(take(cycle(3:-1:1), 7))
     let r = repeated(3.14159)
         @test Iterators.reverse(r) === r
     end
-    let t = (2,3,5,7,11)
+    for t in [(), (1,), (2, 3, 5, 7, 11), NamedTuple(), (a=1,), (a=2, b=3, c=5, d=7, e=11)]
```
Comment (bkamins) [inline test/iterators.jl:698]:
```suggestion
    for t in [(1,), (2, 3, 5, 7, 11), (a=1,), (a=2, b=3, c=5, d=7, e=11)]
```

**Comment by vtjnash** [comment]
Maybe we could consider making this a good default definition too?

**Comment by bkamins** [comment]
> Maybe we could consider making this a good default definition too?

This would assume that by default the iterator supports 1-based indexing.

**Comment by bkamins** [comment]
@JeffBezanson - is this good to merge? Independent of the other discussions about handling of `Iterators.Reverse` in general this definition should be safe to merge I think.

**Review by oscardssmith — APPROVED** [review]

**Pull request merged**